### ISSUE TITLE
Websockets, Google Chrome

### ISSUE THREAD

**Issue by dimisa-RUAdList** [gorhill/uBlock#2142]
`http://novosti-n.org/news/read/113169.html`

Not triggered a general rule: |ws*//*10root25.website^

It is necessary to use specific:  $websocket,domain=novosti-n.org
https://hg.adblockplus.org/ruadlist/rev/8a1c35a93465#l1.8
This is a very popular socket. I would like to block it using the general rule.

Google Chrome 54.0.2840.99 m (64-bit)
uBO 1.9.16, EasyList, RU AdList

**Comment by Snapy** [comment]
Use the [uBO Websocket companion extension](https://chrome.google.com/webstore/detail/ublock-origin-websocket/pgdnlhfefecpicbbihgmbmffkjpaplco).

**Comment by gorhill** [comment]
uBO requires [uBO-WebSocket](https://chrome.google.com/webstore/detail/ublock-origin-websocket/pgdnlhfefecpicbbihgmbmffkjpaplco) to block websocket connections on Chromium-based browsers. Was it installed when you ran your test?

In any case, I did not see any websocket connections in the Network_ dev pane when visiting `http://novosti-n.org/news/read/113169.html`.

**Comment by dimisa-RUAdList** [comment]
I know about uBO-WebSocket. Adblock Plus blocks such things without additional extensions. I am uncomfortable to introduce specific rules on subscription RU AdList individually, only for uBO. Now there is a rule: $websocket,domain=novosti-n.org. But without it, there is such a picture: http://imgur.com/a/HQkrR

**Comment by CrisBRM** [comment]
@dimisa-RUAdList 
>Adblock Plus blocks such things without additional extensions

Are you fucking kidding me? This is your best argument? Why would you want gorhill to bloat uBlock because Chromium's API requires a wrapper?

**Comment by gorhill** [comment]
I just tried these three filters to test with ABP:

    |ws*://*luxupcdna.com^
    |ws*://*adlabs.ru^
    |ws*//*10root25.website^

And no websocket connections were blocked. The site use the workaround to bypass websocket wrapper.

In short, without `$websocket,domain=novosti-n.org`, both ABP and uBO are bypassed.

**Comment by dimisa-RUAdList** [comment]
https://vid.me/kvHr

**Comment by gorhill** [comment]
**TL;DR:** there is no issue when uBO's companion [uBO-WebSocket](https://chrome.google.com/webstore/detail/ublock-origin-websocket/pgdnlhfefecpicbbihgmbmffkjpaplco) is installed.

I had to connect to a Russian VPN server to get websocket connections.

Now the problem with reproducing your specific test case in opening comment is that I needed a RUAdList without your changes in OP. I disabled RUAdlist, then I copied the whole content of RUAdlist in a text file, removed `novosti-n.org` you added, and pasted the result in ABP's raw input in _Custom filters_ pane, then clicked _Apply changes_. I had to wait long minutes before finally ABP parsed the filters, with the browser alerting at regular interval that a script was taking too long.

Now finally I could reproduce the websocket connection to `10root25.website`. Adding `|ws*//*10root25.website^` to ABP's custom filters prevented the connection from being made, so I confirm it works with ABP.

I then proceeded to do exactly the same with uBO, except that I enabled the companion extension and the result was exactly the same (except that uBO parsed the pasted filters in less than a second), uBO was able to block websocket connection attempts to `10root25.website`:

![c](https://cloud.githubusercontent.com/assets/585534/20221556/c0a1cf20-a7ff-11e6-8074-71b54ee3be2f.png)

What this tells is that there is no bypass attempt using WebWorker/Blob for that one specific websocket connections -- thus both ABP's and uBO's wrapper work as expected.

However there were WebWorker/Blob bypass in my other scenario for both ABP/uBO for the websocket connections made by the site to `luxupcdna.com` and `adlabs.ru`, so your filter `$websocket,domain=novosti-n.org` also benefited ABP in that case, but it's up to you if you prefer that someone else than _me_ report this. I think if a site as a history of wanting to work around blocker through websockets, a filter such as `$websocket,domain=novosti-n.org` is certainly not going to be a luxury.

So the issue comes down to this: your filter `|ws*//*10root25.website^` works fine for uBO when its companion extension uBO-WebSocket is also installed.

Now, regarding _"Adblock Plus blocks such things without additional extensions"_:

My reasons to not merge the companion extension with uBO's explained in [that comment](https://github.com/gorhill/uBlock/issues/1801#issuecomment-231601462). ABP's wrapper is no different, it also requires **unconditional** injection of wrapper code in every page/frame. If ABP devs are comfortable with this, it's their choice. Personally I am not.

I know you will very often advise users of your list to install a companion user script to take case of what RUAdList alone cannot take care (it's even a [pinned post on EasyList forum](https://forums.lanik.us/viewforum.php?f=102)), so I don't think it would be out of line that you could also suggest to a uBO user to install uBO's companion uBO-WebSocket where doing so would answer the user's issue.

To be fair you should also consider that there are also many cases which can be properly handled by uBO which does not require a separate user script, so those users not complaining as a result of using uBO rather than ABP.

So to advise uBO users to install uBO-WebSocket when the case arise certainly does not seem worst than being forced to advise ABP users to install a specific user script for cases which ABP can't handle (many of which can be handled by uBO).

**Comment by gorhill** [comment]
Solution: to install install uBO-WebSocket, which is a separate companion, by design.

**Comment by gorhill** [comment]
Off-topic, there is a really serious performance issue with ABP when it comes to manually added filters. It must have been about 20 minutes now ABP is spinning with CPU at 100% for just processing the **removal of all custom filters** (a copy of RUAdlist + minor changes):

![c](https://cloud.githubusercontent.com/assets/585534/20223134/0d5faee4-a806-11e6-8553-3a3fc80dcd6c.png)

I can't make sense of this, the removal of all custom filters means there is **no filter** to parse, and yet ABP is taking even longer to process the operation than it did when I added via copy-paste these filters from RuAdlist (which was still a minutes-long wait).

**Comment by uBlock-user** [comment]
We won't need that companion extension if this [patch](https://bugs.chromium.org/p/chromium/issues/detail?id=129353#c77) gets implemented in Chromium right ?

**Comment by gorhill** [comment]
> We won't need uBO-WebSocket companion extension if this patch gets implemented in Chromium right ?

Correct.

**Comment by pcworld** [comment]
It looks like the patch [has landed](https://bugs.chromium.org/p/chromium/issues/detail?id=129353#c85) in Chromium. Will changes in uBlock be necessary to support this?
